pkg/frontend/querymiddleware/spin_off_subqueries_queryable.go

```diff
@@ -60,6 +74,57 @@ type spinOffSubqueriesQuerier struct {
 
 	// Keep track of response headers received when running embedded queries.
 	responseHeaders *responseHeadersTracker
+	queryCache      map[string]storage.SeriesSet
+	queryCacheMu    *sync.RWMutex
+	inProgress      map[string]chan struct{}
+	inProgressMu    *sync.Mutex
+}
+
+// loadFromCacheOrWait loads a query result from cache or waits for an in-progress query to complete.
+// If the query is not in cache and not in progress, it returns nil.
+func (q *spinOffSubqueriesQuerier) loadFromCacheOrWait(ctx context.Context, key string) storage.SeriesSet {
```

Comment:
Might be worth adding a feature toggle to disable caching and always evaluate the result  every time, just in case we run into any bugs with this.